File: src/main/kotlin/com/compiler/max_args.kt
package com.compiler

fun maxArgs(statement: Statement): Int =
    when (statement) {
        is Print -> statement.expressions.size
        is Assignment -> statement.findPrint()?.let { print -> maxArgs(print) } ?: 0
        is CompoundStatement ->
            maxOf(maxArgs(statement.statement1), maxArgs(statement.statement2))
    }

private fun Assignment.findPrint(): Print? =
    when (expression) {
        is StatementThenExpression ->
            when (expression.statement) {
                is Print -> expression.statement
                else -> null
            }
        else -> null
    }

File: ocaml/compiler/lib/maxargs.ml
open Grammar
open Math 

exception Todo

let rec max_args (statement) =
  match statement with
  | Print expressions -> (
      let args_for_expressions = List.map args_for expressions in
      (max (List.length expressions) (max_of args_for_expressions))
    )
  | Assignment (_, expression) -> args_for expression
  | CompoundStatement (_, _) -> raise Todo
  and args_for (expression) =
    match expression with
    | StatementThenExpression (statement, _) -> max_args statement
    | _ -> 0
